crates/services/parallel-executor/src/scheduler.rs

```diff
@@ -0,0 +1,1179 @@
+    ) -> Result<(), SchedulerError> {
+        self.tx_left = self.tx_left.checked_sub(tx_number_to_add).ok_or(
+            SchedulerError::InternalError(
+                "Cannot add more transactions: tx_left underflow".to_string(),
+            ),
+        )?;
+        self.tx_size_left = self.tx_size_left.checked_sub(tx_size_to_add).ok_or(
+            SchedulerError::InternalError(
+                "Cannot add more transactions: tx_size_left underflow".to_string(),
+            ),
+        )?;
+        self.gas_left = self.gas_left.checked_sub(gas_to_add).ok_or(
+            SchedulerError::InternalError(
+                "Cannot add more transactions: gas_left underflow".to_string(),
+            ),
+        )?;
+        Ok(())
+    }
+
+    fn is_worker_idling(&self) -> bool {
+        !self.current_available_workers.is_empty()
+            && self.state == SchedulerState::TransactionsReadyForPickup
+    }
+
+    fn new_executable_transactions(&mut self) {
+        self.state = SchedulerState::TransactionsReadyForPickup;
+    }
+
+    fn ask_new_transactions_batch<TxSource: TransactionsSource>(
+        &mut self,
+        tx_source: &mut TxSource,
+        start_execution_time: tokio::time::Instant,
+        initial_gas: u64,
+        total_execution_time: Duration,
+    ) -> Result<PreparedBatch, SchedulerError> {
+        let spent_time = start_execution_time.elapsed();
+        // Time left in percentage to have the gas percentage left
+        // TODO: Maybe avoid as u32
+        let current_gas = std::cmp::min(
+            initial_gas
+                .saturating_mul(
+                    (total_execution_time.as_millis() as u64)
+                        .saturating_sub(spent_time.as_millis() as u64),
+                )
+                .saturating_div(total_execution_time.as_millis() as u64),
+            self.gas_left
+                .saturating_div(self.config.number_of_cores.get() as u64),
+        );
+
+        let (batch, filtered, filter) = tx_source.get_executable_transactions(
+            current_gas,
+            self.tx_left,
+            self.tx_size_left,
+            Filter {
+                excluded_contract_ids: std::mem::take(
+                    &mut self.current_executing_contracts,
+                ),
+            },
+        );
+        self.current_executing_contracts = filter.excluded_contract_ids;
+
+        if batch.is_empty() {
+            if filtered == TransactionFiltered::Filtered {
+                self.state = SchedulerState::WaitingForWorker;
+            } else {
+                self.state = SchedulerState::WaitingForNewTransaction;
+            }
+        }
+
+        let prepared_batch = prepare_transactions_batch(batch)?;
+        self.update_constraints(
+            prepared_batch.number_of_transactions,
+            prepared_batch.total_size,
+            prepared_batch
+                .gas
+                .saturating_div(self.config.number_of_cores.get() as u64)
```

Comment:
Why do we need to divide here? I thought we want to track the total gas left

Reply:
Yeah right we are already dividing above to get the gas. The problem is that we are also dividing the gas for blobs which should consume gas for all cores so I added a multiply for the blobs. If it's not clear we can discuss it.